### PR TITLE
feat: track store r/w consume for enabling plain store

### PR DESCRIPTION
**Pull request by forcodedancing**
### Description

This pr will track store r/w consume into deliver tx result.

### Rationale

For enabling plain store without IAVL, we need to enhance the `LastResultHash` when validating blocks.

### Example

new flag `EnablePlainStore = true`

### Changes

Notable changes:
* new flag

### CODE REVIEW

**Review by guagualvcha — APPROVED** [review]

File: baseapp/abci.go [hunk 1]
```diff
@@ -444,11 +445,23 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDeliv
 		return sdkerrors.ResponseDeliverTxWithEvents(err, gInfo.GasWanted, gInfo.GasUsed, sdk.MarkEventsToIndex(anteEvents, app.indexEvents), app.trace)
 	}
 
+	app.deliverState.ctx.Logger().Debug("Gas info rw used", "RwUsed", gInfo.RwUsed, "height", app.deliverState.ctx.BlockHeight())
+	rwUsedBz := sdk.Uint64ToBigEndian(gInfo.RwUsed)
+
+	msgData := &sdk.TxMsgData{}
+	proto.Unmarshal(result.Data, msgData)
```
Comment (guagualvcha) [inline baseapp/abci.go:452]: can we not do `Unmarshal` and `marshal` here, as it is duplicated calculation, is it possible to achieve it in  `makeABCIData`?

Comment (guagualvcha) [reply baseapp/abci.go:452]: Not only tx need this `RwUsed`, but also begin block  and end block.

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -85,7 +85,6 @@ type BaseApp struct { //nolint: maligned
 	//
 	// checkState is set on InitChain and reset on Commit
 	// deliverState is set on InitChain and BeginBlock and set to nil on Commit
-	// queryState is set on InitChain and BeginBlock
```
Comment (guagualvcha) [inline baseapp/baseapp.go:88]: why delete this?

File: baseapp/baseapp.go [hunk 3]
```diff
@@ -168,6 +167,9 @@ type BaseApp struct { //nolint: maligned
 
 	// enableUnsafeQuery defines whether the unsafe queries will be enabled or not
 	enableUnsafeQuery bool
+
+	// enablePlainStore defines whether uses plain db store type or not
+	enablePlainStore bool
```
Comment (guagualvcha) [inline baseapp/baseapp.go:172]: `enablePlainStore` means the `skipAppHashVerify` should be true, any way to verify they are consistent?

Comment (guagualvcha) [reply baseapp/baseapp.go:172]: also highlight the impaction of this configure.

File: baseapp/baseapp.go [hunk 4]
```diff
@@ -290,7 +292,7 @@ func (app *BaseApp) MountStores(keys ...storetypes.StoreKey) {
 	for _, key := range keys {
 		switch key.(type) {
 		case *storetypes.KVStoreKey:
-			if !app.fauxMerkleMode {
+			if !app.fauxMerkleMode && !app.enablePlainStore {
```
Comment (guagualvcha) [inline baseapp/baseapp.go:295]: please use `IsIavlStore` instead?

File: baseapp/baseapp.go [hunk 5]
```diff
@@ -314,7 +316,7 @@ func (app *BaseApp) MountStores(keys ...storetypes.StoreKey) {
 // BaseApp multistore.
 func (app *BaseApp) MountKVStores(keys map[string]*storetypes.KVStoreKey) {
 	for _, key := range keys {
-		if !app.fauxMerkleMode {
+		if !app.fauxMerkleMode && !app.enablePlainStore {
```
Comment (guagualvcha) [inline baseapp/baseapp.go:319]: use `IsIavlStore` instead?

File: types/context.go [hunk 6]
```diff
@@ -313,15 +314,13 @@ func (c Context) Value(key interface{}) interface{} {
 // ----------------------------------------------------------------------------
 
 // KVStore fetches a KVStore from the MultiStore.
-// Remove gas metering
 func (c Context) KVStore(key storetypes.StoreKey) storetypes.KVStore {
-	return c.MultiStore().GetKVStore(key)
+	return gaskv.NewStore(c.ms.GetKVStore(key), c.gasMeter, c.kvGasConfig)
```
Comment (guagualvcha) [inline types/context.go:318]: I prefer `	return gaskv.NewStore(c.MultiStore().GetKVStore(key), c.gasMeter, c.kvGasConfig)`

File: types/context.go [hunk 7]
```diff
@@ -313,15 +314,13 @@ func (c Context) Value(key interface{}) interface{} {
...
 }
 
 // TransientStore fetches a TransientStore from the MultiStore.
-// Remove gas metering
 func (c Context) TransientStore(key storetypes.StoreKey) storetypes.KVStore {
-	return c.MultiStore().GetKVStore(key)
+	return gaskv.NewStore(c.ms.GetKVStore(key), c.gasMeter, c.kvGasConfig)
```
Comment (guagualvcha) [inline types/context.go:323]: prefer `return gaskv.NewStore(c.MultiStore().GetKVStore(key), c.gasMeter, c.kvGasConfig)`

**Review by guagualvcha — CHANGES_REQUESTED** [review]
need add more gasrw for endblock and begin block.

File: baseapp/baseapp.go (hunk 2 above)
Comment (forcodedancing) [reply baseapp/baseapp.go:88]: It is duplicated comment. There is same comment in the following line.

File: types/context.go (hunk 7 above)
Comment (forcodedancing) [reply types/context.go]: updated

File: types/context.go (hunk 6 above)
Comment (forcodedancing) [reply types/context.go]: updated

File: baseapp/baseapp.go (hunk 5 above)
Comment (forcodedancing) [reply baseapp/baseapp.go]: updated

File: baseapp/baseapp.go (hunk 4 above)
Comment (forcodedancing) [reply baseapp/baseapp.go]: updated

File: baseapp/baseapp.go (hunk 3 above)
Comment (forcodedancing) [reply baseapp/baseapp.go:172]: added some comments to app.toml

File: baseapp/abci.go [hunk 8]
```diff
@@ -1166,6 +1169,7 @@ func SplitABCIQueryPath(requestPath string) (path []string) {
 func (app *BaseApp) getContextForProposal(ctx sdk.Context, height int64) sdk.Context {
 	if height == app.initialHeight {
 		ctx, _ = app.deliverState.ctx.CacheContext()
+		ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
```
Comment (guagualvcha) [inline baseapp/abci.go:1172]: what kind of `GasMeter` are we using during the block execution? if use `basicGasMeter`, may cause `ErrorOutOfGas`

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**